### PR TITLE
perl-bio-ensembl-io: new package

### PR DESCRIPTION
**Pull request by teaguesterling**
Add the `perl-bio-ensembl-io` corresponding to the `Bio::Ensembl::IO` Perl module. This has dependencies that have been added in separate PRs. This shouldn't be merged until #44505 and #44413 are merged in.

Note about the install approach: this is a Perl module but it doesn't have appropriate installer. In VEP it's downloaded and moved into place by the bundled install.PL script. This is an attempt to make it a bit more conventional.

### CODE REVIEW

**Comment by EbiArnie** [comment]
Please remove Bio::DB::Bigfile and add a dependency to perl-bio-ensembl.

**Comment by EbiArnie** [comment]
Apart from that, looks good to me.

**Comment by teaguesterling** [comment]
> Apart from that, looks good to me.

Bio::DB::Bigfile is still here as the audit checks fail without it. Once #44505 is merged, this will only be for `perl-bio-ensembl-io`.

**Review by EbiArnie — CHANGES_REQUESTED** [review]
perl-bio-bigfile is now merged. Please remove it from this PR.

**Comment by EbiArnie** [comment]
So, this is a bit awkward.

Your package is fine with respect to the cpanfile.
But I tried to run the test suite and ran into quite a few issues.
It turns out that the cpanfile is missing many dependencies.
These are installed by the other parts of the API, e.g. ensembl-variation. Our installation guide expects everything to be set up together. But this is not strictly necessary, ensembl-io is perfectly usable on its own - with the correct dependencies, of course.
See below.
I work at the EBI, so I will try to get that updated as soon as possible.
The changes should show up in the main branch for ensembl-io soon enough, but I have to see what I can do for the released versions.

Meanwhile, I would suggest to go with these dependencies:

```
requires 'Bio::DB::BigFile';
requires 'Bio::DB::HTS';
requires 'Bio::DB::Sam';
requires 'Bio::Tools::CodonTable';
requires 'Compress::Bzip2';
requires 'JSON';
requires 'Try::Tiny';
requires 'URI::Escape';
```

**Review by EbiArnie — CHANGES_REQUESTED** [review]
Please update dependencies as described in previous comment.

**Comment by teaguesterling** [comment]
Not awkward at all! This is great. I wanted to create this to make setting up VEP in air gapped environments easier. Catching these things in advance will only make life easier in the future.

**Comment by teaguesterling** [comment]
@EbiArnie can you let me know how you tested so I can make sure I cover the same?

**Comment by EbiArnie** [comment]
I run the test suite that comes with the repo. I would look at the travis CI test stuff for how to set things up.

For me, it goes like this:
1. Create Spack env with the new perl-ensembl-io, install and activate.
2. Add (at least) the ensembl-test and ensembl-variation modules to PERL5LIB (including their dependencies).
3. I have a local MySQL DB. There is a file `modules/t/MultiTestDB.conf` which configures the tests to use this DB. I usually copy the one from the travis test. SQLite should also work.
4. Set up travis user and give it grants on the DB.
5. Now look at harness.sh from the travis dir. I do not need the coverage reports, so it's `perl ${repo-dir}/ensembl-test/scripts/runtests.pl modules/t`.

**Comment by EbiArnie** [comment]
I was able to run the test suite successfully with these dependencies:

```
    depends_on("perl-bio-bigfile")
    depends_on("perl-bio-db-hts")
    depends_on("perl-bio-ensembl")
    depends_on("perl-bioperl")
    depends_on("perl-json")
    depends_on("perl-try-tiny")
    depends_on("perl-uri")
    depends_on("vcftools")
#    depends_on("perl-bio-samtools")
#    depends_on("perl-compress-bzip2")
```

The deps for samtools and compress-bzip2, I believe they are needed, but not covered by the test suite.
Also, they are not yet in Spack, so I excluded them for now.

To run the tests, you need ensembl-test and that needs many more modules. But since this is not the regular Spack package installation phase that runs the tests, I feel we should exclude them for now. 

One more thing. There should be a version constraint for "ensembl-io release 112 needs ensembl release 112".

**Comment by teaguesterling** [comment]
@spackbot fix style

**Comment by teaguesterling** [comment]
@spackbot fix style

**Comment by EbiArnie** [comment]
Sorry to keep you waiting for so long.
I ran the test suite again and it took me a while to find the source for a failure I was seeing.
We need to depend on bioperl 1.6.924. At some point in later versions, bioperl changed a codon table. That leads to subtle differences that the test suite does not like.
Also, Compress::Bzip2 is now merged, so I put that in too.

```
    depends_on("perl-bioperl@1.6.924")
    depends_on("perl-compress-bzip2")
```

With that, I think this is good to go in.

**Comment by teaguesterling** [comment]
> We need to depend on bioperl 1.6.924.

Is it this version exactly, or is there a range? Additionally, do you think that the following would be appropriate?

```
depends_on("perl-bioperl")
depends_on("perl-bioperl@1.6.924", type="test")
```


Also, thank you for what I can only imagine was a gurgling debugging session to sort that out.

**Comment by EbiArnie** [comment]
You're welcome.

We really need bioperl 1.6.924. Or, more precisely, we need a certain variant of `Bio::Tools::CodonTable` and bioperl < 1.7.
So >= 1.6.924, < 1.7 would also be fine, probably. But there is no newer version than 1.6.924 and it's unlikely there will ever be one. Bioperl has moved on for good reason, and we haven't made the switch yet because it breaks a lot of our code and tests.

Having the version fixed only for the test is not enough. Using a different codon table changes the output of our code. I can not judge if the new output would be better or worse, but as I said, it would break things here, so we should keep this at the fixed version for now.

I think it should be:
```
depends_on("perl-bioperl@1.6.924")
```

File: var/spack/repos/builtin/packages/perl-bio-ensembl-io/package.py
Comment (EbiArnie) [inline var/spack/repos/builtin/packages/perl-bio-ensembl-io/package.py]: Please constrain this to version 1.6.924.

**Review by EbiArnie — CHANGES_REQUESTED** [review]

**Comment by teaguesterling** [comment]
It looks like we may need to add that version of BioPerl to to the `perl-bioperl` package as well. Happy to do that as another PR later today. 

```
PKG-DIRECTIVES: 1 issue found
1. perl-bio-ensembl-io: dependency on perl-bioperl@1.6.924 cannot be satisfied by known versions of perl-bioperl
    happening in /home/runner/work/spack/spack/var/spack/repos/builtin/packages/perl-bio-ensembl-io/package.py
    known versions of perl-bioperl are 1.7.6, 1.007002
```

File: var/spack/repos/builtin/packages/perl-bio-ensembl-io/package.py
Comment (teaguesterling) [reply var/spack/repos/builtin/packages/perl-bio-ensembl-io/package.py]: Done pending #46213

**Comment by teaguesterling** [comment]
@EbiArnie, this should be good to go now, assuming the checks pass.

**Comment by EbiArnie** [comment]
The test suite runs fine now - good to go.

**Review by EbiArnie — APPROVED** [review]
OK

**Pull request merged**
